js/piwik.js

```diff
@@ -5571,7 +5596,55 @@ if (typeof window.Matomo !== 'object') {
              * @param generationTime
              */
             this.setGenerationTimeMs = function(generationTime) {
-                logConsoleError('setGenerationTimeMs is no longer supported since Matomo 4. The call will be ignored. There is currently no replacement yet.');
+                logConsoleError('setGenerationTimeMs is no longer supported since Matomo 4. The call will be ignored. The replacement is setPagePerformanceTiming.');
+            };
+
+            /**
+             * Replace setGenerationTimeMs with this more generic function
+             * Use in SPA
+             * @param networkTimeInMs
+             * @param serverTimeInMs
+             * @param transferTimeInMs
+             * @param domProcessingTimeInMs
+             * @param domCompletionTimeInMs
+             * @param onloadTimeInMs
+             */
+            this.setPagePerformanceTiming = function(
+                networkTimeInMs, serverTimeInMs, transferTimeInMs,
+                domProcessingTimeInMs, domCompletionTimeInMs, onloadTimeInMs
+            ) {
+                /*property pf_net, pf_srv, pf_tfr, pf_dm1, pf_dm2, pf_onl */
+                var data = {
+                    pf_net: networkTimeInMs,
+                    pf_srv: serverTimeInMs,
+                    pf_tfr: transferTimeInMs,
+                    pf_dm1: domProcessingTimeInMs,
+                    pf_dm2: domCompletionTimeInMs,
+                    pf_onl: onloadTimeInMs
+                }, errorCaught = false, message;
+
+                try {
+                    configPagePerformanceTiming = queryStringify(data, isNumber, true);
+                } catch (error) {
+                    errorCaught = true;
+                    message = error.toString() + ' ';
+                }
+                if (configPagePerformanceTiming !== '' && !errorCaught) {
+                    performanceTracked = false; // to ensure the values are sent (again)
+                    performanceAvailable = true; // so appendAvailablePerformanceMetrics will be called directly
+                    // Otherwise performanceAvailable will be set when the pageload finished, but there is no need
+                    // to wait for that, when the values are set manually.
+                } else {
+                    if (errorCaught) {
+                        message = 'Please ensure to only supply numbers for each parameter to ' +
+                            'setPagePerformanceTiming().';
+                    } else {
+                        message = 'setPagePerformanceTiming() called without parameters. It only makes sense to call ' +
+                            'this function with at least one performance parameter like networkTimeInMs. Also, ' +
+                            'please ensure to only supply numbers for each parameter.';
+                    }
+                    logConsoleError(message);
+                }
```

Comment:
This code has a bit more complexity as needed. Wouldn't something like this be a bit simpler to read:

```js
var data = {
    pf_net: networkTimeInMs,
    pf_srv: serverTimeInMs,
    pf_tfr: transferTimeInMs,
    pf_dm1: domProcessingTimeInMs,
    pf_dm2: domCompletionTimeInMs,
    pf_onl: onloadTimeInMs
};

try {
    configPagePerformanceTiming = queryStringify(data, isNumber, true);
    if (configPagePerformanceTiming === '') {
        logConsoleError('setPagePerformanceTiming() called without parameters. It only makes sense to call ' +
            'this function with at least one performance parameter like networkTimeInMs. Also, ' +
            'please ensure to only supply numbers for each parameter.');
        return;
    }
                    
    performanceTracked = false; // to ensure the values are sent (again)
    performanceAvailable = true; // so appendAvailablePerformanceMetrics will be called directly
    // Otherwise performanceAvailable will be set when the pageload finished, but there is no need
    // to wait for that, when the values are set manually.
} catch (error) {
    logConsoleError(error.toString() + ' Please ensure to only supply numbers for each parameter to ' +
        'setPagePerformanceTiming().');
}
```

Reply:
Agreed. Thanks!